src/metabase/models/task_history.clj

```diff
@@ -71,45 +87,54 @@
 (def ^:private TaskHistoryInfo
   "Schema for `info` passed to the `with-task-history` macro."
   [:map {:closed true}
-   [:task                          ms/NonBlankString] ; task name, i.e. `send-pulses`. Conventionally lisp-cased
-   [:db_id        {:optional true} [:maybe :int]]     ; DB involved, for sync operations or other tasks where this is applicable.
-   [:task_details {:optional true} [:maybe :map]]])   ; additional map of details to include in the recorded row
+   [:task                             ms/NonBlankString] ; task name, i.e. `send-pulses`. Conventionally lisp-cased
+   [:db_id           {:optional true} [:maybe :int]]     ; DB involved, for sync operations or other tasks where this is applicable.
+   ;; a function that takes the result of the task and returns a map of additional info to update task history when the task succeeds
+   [:on-success-info {:optional true} [:maybe [:=> [:cat :any] :map]]]
+   [:task_details    {:optional true} [:maybe :map]]])   ; additional map of details to include in the recorded row
 
-(defn- save-task-history! [start-time-ms info]
-  (let [end-time-ms (System/currentTimeMillis)
-        duration-ms (- end-time-ms start-time-ms)]
-    (try
-      (first (t2/insert-returning-instances! TaskHistory
-                                             (assoc info
-                                                    :started_at (t/instant start-time-ms)
-                                                    :ended_at   (t/instant end-time-ms)
-                                                    :duration   duration-ms)))
-      (catch Throwable e
-        (log/warn e "Error saving task history")))))
+(def ^:private ns->ms #(int (/ % 1e6)))
```

Comment:
An even better way to DRY stuff up would be to elevate [an existing helper](https://github.com/metabase/metabase/blob/e8fc1a0a60a4c2e276866b3b135474f04d98fe6f/src/metabase/upload.clj#L469) from the upload module.

Definitely out of scope, but timers in serialization, liquibase, and the query-processor cache code could also use this helper.

Probably worth renaming the one arg to be more more explicit, and here are some suggested docstrings. The idea is to fully encapsulate the fact that we use `System/nanoTime`

```clojure
(defn start-timer 
  "Return a timer tracking from when this was called, for use with [[since-ms]]."
  [] 
  (System/nanoTime))

(defn since-ms 
  "Calculate the number of milliseconds since the given timer was started."
  [timer-started-ns]
   (/ (- (System/nanoTime) timer-started-ns) 1e6))
```

Reply:
awesome, there are a lots of place that can reuse this so I'll have a follow up for this!

Reply:
Util should probably have a link to something explaining why nanoTime is better than currentMillis, since it's so tempting to swap back to remove the division.